torch_xla/csrc/runtime/ifrt_computation_client.cc

```diff
@@ -0,0 +1,643 @@
+        ifrt_data->buffer
+            ->DisassembleIntoSingleDeviceArrays(
+                xla::ifrt::ArrayCopySemantics::kAlwaysCopy)
+            .value();
+
+    for (auto array : arrays) {
+      shards.push_back(std::make_shared<IfrtData>(
+          PjRtDeviceToString(array->sharding().devices()[0]), array));
+    }
+  } else {
+    shards.push_back(data);
+  }
+  return shards;
+}
+
+ComputationClient::DataPtr IfrtComputationClient::GetDataShard(
+    ComputationClient::DataPtr data, size_t index) {
+  tsl::profiler::TraceMe activity("IfrtComputationClient::GetDataShard",
+                                  tsl::profiler::TraceMeLevel::kInfo);
+  return GetDataShards(data)[index];
+}
+
+ComputationClient::DataPtr IfrtComputationClient::WrapDataShards(
+    const std::vector<DataPtr>& shards, std::string device, xla::Shape shape,
+    xla::OpSharding sharding) {
+  // TODO: implement CreateDataPlaceholder for sharded data
+  if (shards.size() == 0) {
+    TF_LOG(INFO) << "creating sharded placeholder";
+    return std::make_shared<IfrtData>(
+        device, shape, tsl::RCReference<xla::ifrt::Array>(), sharding);
+  }
+  std::vector<tsl::RCReference<xla::ifrt::Array>> arrays;
+  std::vector<xla::ifrt::Shape> shard_shapes;
+  for (auto& shard : shards) {
+    auto ifrt_shard = std::dynamic_pointer_cast<IfrtData>(shard);
+    arrays.push_back(ifrt_shard->buffer);
+    shard_shapes.push_back(ifrt_shard->buffer->shape());
+  }
+  xla::ifrt::Shape ifrt_shape(shape.dimensions());
+  xla::ifrt::DeviceList devices_list({client_->addressable_devices().begin(),
+                                      client_->addressable_devices().end()});
+  XLA_CHECK_EQ(shard_shapes.size(), devices_list.size());
+  std::unique_ptr<xla::ifrt::Sharding> ifrt_sharding =
+      xla::ifrt::ConcreteSharding::Create(devices_list, xla::ifrt::MemoryKind(),
+                                          ifrt_shape, shard_shapes);
+  // TODO: Attach HloSharding instead when it is supported
+  // std::unique_ptr<xla::ifrt::Sharding> ifrt_sharding =
+  // xla::ifrt::HloSharding::Create(
+  //   devices_list,
+  //   xla::ifrt::MemoryKind(),
+  //   xla::HloSharding::FromProto(sharding).value()
+  // );
+  tsl::RCReference<xla::ifrt::Array> sharded_array =
+      client_
+          ->AssembleArrayFromSingleDeviceArrays(
+              ifrt_shape, std::move(ifrt_sharding), absl::MakeSpan(arrays),
+              xla::ifrt::ArrayCopySemantics::kAlwaysCopy)
+          .value();
+  return std::make_shared<IfrtData>(device, shape, sharded_array, sharding);
+}
+
+std::optional<xla::OpSharding> IfrtComputationClient::GetDataSharding(
+    DataPtr handle) {
+  auto ifrt_data = std::dynamic_pointer_cast<IfrtData>(handle);
+  return ifrt_data->sharding_;
+}
+
+std::vector<ComputationClient::DataPtr> IfrtComputationClient::TransferToServer(
+    absl::Span<const std::shared_ptr<const TensorSource>> tensors) {
+  metrics::TimedSection timed(TransferToServerMetric());
```

Comment:
is transfer async? If so this metric won't capture the real device transfer time right?

Reply:
Fixed